Random inherited from person in `Lifestyle.on_birth` method can sometimes be child itself

In the following line

https://github.com/UCL/TLOmodel/blob/a8a0af0e75f55bd0894c4c98875a53d347912c5b/src/tlo/methods/enhanced_lifestyle.py#L793

a person ID index is randomly selected from the alive individuals on births where the `mother_id` for the child is equal to -1, with the intention I think for this to then be a replacement 'mother' individual to inherit lifestyle properties from. As at the point this lines run the `is_alive` property for the child who's birth triggered the `on_birth` call is `True` there is a small but non-zero probability that the randomly generated ID will correspond to the child itself. If this happens then the various properties inherited using the `_id_inherit_from` index do not get initialised correctly as they are simply reset to there current values.

This came up when running the tests across multiple seeds, with there being a failure in `tests/test_module_dependencies.py::test_module_dependencies_complete` for the `Contraception` module when run with seed 3512589365 with error
```
_________ test_module_dependencies_complete[3512589365-Contraception] __________
tests/test_module_dependencies.py:222: in test_module_dependencies_complete
    register_modules_and_simulate(
tests/test_module_dependencies.py:83: in register_modules_and_simulate
    sim.simulate(end_date=simulation_end_date)
.tox/py38/lib/python3.8/site-packages/tlo/simulation.py:235: in simulate
    self.fire_single_event(event, date)
.tox/py38/lib/python3.8/site-packages/tlo/simulation.py:275: in fire_single_event
    event.run()
.tox/py38/lib/python3.8/site-packages/tlo/events.py:34: in run
    self.apply(self.target)
.tox/py38/lib/python3.8/site-packages/tlo/methods/stunting.py:418: in apply
    idx_will_be_stunted = self.apply_model(
.tox/py38/lib/python3.8/site-packages/tlo/methods/stunting.py:461: in apply_model
    annual_prob = model.predict(df.loc[mask]).clip(upper=1.0)
.tox/py38/lib/python3.8/site-packages/tlo/lm.py:417: in predict
    column_resolvers = self._get_column_resolvers(df, **kwargs)
.tox/py38/lib/python3.8/site-packages/tlo/lm.py:374: in _get_column_resolvers
    column_resolvers[cleaned_name] = col.astype(dtype)
.tox/py38/lib/python3.8/site-packages/pandas/core/generic.py:5874: in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
.tox/py38/lib/python3.8/site-packages/pandas/core/internals/managers.py:631: in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
.tox/py38/lib/python3.8/site-packages/pandas/core/internals/managers.py:427: in apply
    applied = getattr(b, f)(**kwargs)
.tox/py38/lib/python3.8/site-packages/pandas/core/internals/blocks.py:648: in astype
    values = self.values.astype(dtype)
.tox/py38/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:422: in astype
    raise ValueError("Cannot convert float NaN to integer")
E   ValueError: Cannot convert float NaN to integer
```

On debugging I found that the relevant column which was causing the error in the `column_resolvers[cleaned_name] = col.astype(dtype)` line was the `li_weath` column, which contained a single `NaN` value for the last individual in the population (person ID 1000). After putting a breakpoing in the `Lifestyle.on_birth` call I found that `_id_inherit_from` was getting set to 1000.

--- a/src/tlo/methods/demography.py
+++ b/src/tlo/methods/demography.py
@@ -19,7 +19,7 @@ from tlo.methods.causes import (
     collect_causes_from_disease_modules,
     create_mappers_from_causes_to_label,
 )
-from tlo.util import create_age_range_lookup
+from tlo.util import create_age_range_lookup, get_person_id_to_inherit_from
 
 # Standard logger
 logger = logging.getLogger(__name__)
@@ -280,8 +280,9 @@ class Demography(Module):
 
         fraction_of_births_male = self.parameters['fraction_of_births_male'][self.sim.date.year]
 
-        # Determine characteristics that are inherited from mother (and if no mother, from a randomly selected person)
-        _id_inherit_from = mother_id if mother_id != -1 else rng.choice(df.index[df.is_alive])
+        # Determine characteristics that are inherited from mother (and if no mother,
+        # from a randomly selected person)
+        _id_inherit_from = get_person_id_to_inherit_from(child_id, mother_id, df, rng)
         _district_num_of_residence = df.at[_id_inherit_from, 'district_num_of_residence']
         _district_of_residence = df.at[_id_inherit_from, 'district_of_residence']
         _region_of_residence = df.at[_id_inherit_from, 'region_of_residence']

--- a/src/tlo/methods/enhanced_lifestyle.py
+++ b/src/tlo/methods/enhanced_lifestyle.py
@@ -10,6 +10,7 @@ import pandas as pd
 
 from tlo import DateOffset, Module, Parameter, Property, Types, logging
 from tlo.events import PopulationScopeEventMixin, RegularEvent
+from tlo.util import get_person_id_to_inherit_from
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
@@ -789,11 +790,9 @@ class Lifestyle(Module):
 
         # Determine id from which characteristics that inherited (from mother, or if no
         # mother, from a randomly selected alive person that is not child themself)
-        if mother_id == -1:
-            alive_persons_not_including_child = df.index[df.is_alive].drop(child_id)
-            _id_inherit_from = self.rng.choice(alive_persons_not_including_child)
-        else:
-            _id_inherit_from = mother_id
+        _id_inherit_from = get_person_id_to_inherit_from(
+            child_id, mother_id, df, self.rng
+        )
 
         df.at[child_id, 'li_urban'] = df.at[_id_inherit_from, 'li_urban']
         df.at[child_id, 'li_wealth'] = df.at[_id_inherit_from, 'li_wealth']

--- a/src/tlo/util.py
+++ b/src/tlo/util.py
@@ -407,3 +407,22 @@ def hash_dataframe(dataframe: pd.DataFrame):
         return df.applymap(lambda x: tuple(x) if isinstance(x, list) else x)
 
     return hashlib.sha1(pd.util.hash_pandas_object(coerce_lists_to_tuples(dataframe)).values).hexdigest()
+
+
+def get_person_id_to_inherit_from(child_id, mother_id, population_dataframe, rng):
+    """Get index of person to inherit properties from.
+
+    Should be specified mother_id unless this is equal to -1 (for example for
+    individuals generated at population initialisation with no mother set) in which
+    case an individual from currently alive persons, _excluding the person the index to
+    inherit from is being computed for_, is randomly chosen.
+    """
+    if mother_id == -1:
+        # Get indices of alive persons and try to drop child_id from these indices if
+        # present, ignoring any errors if child_id not currently in population dataframe
+        alive_persons_not_including_child = population_dataframe.index[
+            population_dataframe.is_alive
+        ].drop(child_id, errors="ignore")
+        return rng.choice(alive_persons_not_including_child)
+    else:
+        return mother_id

--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -222,3 +222,30 @@ def test_logs_parsing(tmpdir):
 
         # check the created pickle file is not empty
         assert os.path.getsize(path_to_tmpdir / f"{key}.pickle") != 0
+
+
+def test_get_person_id_to_inherit_from(rng: np.random.RandomState):
+    population_size = 1000
+    num_test = 5
+    for child_id, mother_id in rng.randint(0, population_size, size=(num_test, 2)):
+        # population_dataframe and rng arguments should be unused if mother_id != -1
+        assert mother_id == tlo.util.get_person_id_to_inherit_from(
+            child_id, mother_id, population_dataframe=None, rng=None
+        )
+    population_dataframe = pd.DataFrame(
+        {
+            "is_alive": rng.choice((True, False), size=population_size),
+            "sex": rng.choice(("F", "M"), size=population_size),
+            "age_years": rng.randint(0, 100, size=population_size),
+        }
+    )
+    mother_id = -1
+    for child_id in rng.choice(
+        population_dataframe.index[population_dataframe.is_alive], size=num_test
+    ):
+        inherit_from_id = tlo.util.get_person_id_to_inherit_from(
+            child_id, mother_id, population_dataframe, rng
+        )
+        assert inherit_from_id != mother_id
+        assert inherit_from_id != child_id
+        assert population_dataframe.loc[inherit_from_id].is_alive

